File: Sources/COCHelperCore/CapitalRaidRowCache.swift
import Foundation

/// Issue #221：突袭周末 row identity 生命周期缓存（ephemeral render sidecar）。
///
/// 在 #211 轻量 `tripleKey#seq` 之上补全 UI 行 identity 生命周期：
/// - load more 增量 append，旧行 ID 不变；
/// - refresh 仅做可证明匹配并更新 payload；
/// - duplicate/reorder 歧义时 bump generation 并 fail-closed 重建；
/// - View 只读取已维护好的 `rows`，不在 render path 重算或深比较 page。
public final class CapitalRaidRowCache {
    /// 缓存更新语义（由 AppModel 在状态转换点调用）。
    public enum Update: Sendable {
        /// 冷启动 / 持久化 lastGood 恢复：新 generation，不做 previous 匹配。
        case initial(page: OfficialCapitalRaidPage)
        /// 首屏刷新成功（同 parser 版本）：尝试 reconcile，歧义则 reset。
        case refreshSuccess(page: OfficialCapitalRaidPage)
        /// 加载更多成功：prefix 匹配则 append；overlap 歧义时 reset。
        case loadMoreSuccess(
            page: OfficialCapitalRaidPage,
            reconciliation: CapitalRaidPaginationMerge.LoadMoreReconciliation = .identityPreserving
        )
        /// 跨 parser 版本重建：明确 bump generation。
        case parserRebuild(page: OfficialCapitalRaidPage)
        /// 请求失败保留 last-good：row state 不变。
        case failureRetain
        /// 清除该 tag 的 ephemeral row state。
        case clear
    }

    public private(set) var generation: UInt64 = 0
    public private(set) var rows: [CapitalRaidSeasonRow] = []
    public private(set) var buildCount = 0

    private var nextSequenceByTriple: [String: Int] = [:]

    public init() {}

    @discardableResult
    public func apply(_ update: Update) -> [CapitalRaidSeasonRow] {
        switch update {
        case .failureRetain:
            return rows
        case .clear:
            generation = 0
            rows = []
            nextSequenceByTriple = [:]
            return rows
        case .initial(let page), .parserRebuild(let page):
            resetAndBuild(from: page.items)
            return rows
        case .refreshSuccess(let page):
            reconcileRefresh(with: page.items)
            return rows
        case .loadMoreSuccess(let page, let reconciliation):
            if reconciliation == .ambiguous {
                resetAndBuild(from: page.items)
            } else {
                reconcileLoadMore(with: page.items)
            }
            return rows
        }
    }

    // MARK: - reconcile

    private func reconcileRefresh(with seasons: [OfficialCapitalRaidSeason]) {
        if rows.isEmpty {
            resetAndBuild(from: seasons)
            return
        }
        if seasons.count < rows.count {
            if seasons.isEmpty {
                resetAndBuild(from: seasons)
                return
            }
            // Issue #230：截短 refresh 必须基于完整旧缓存判断 duplicate 歧义，
            // 不得先用 prefix 丢掉被截掉行的身份证据。
            if canSafelyReconcileTruncatedRefresh(oldRows: rows, newSeasons: seasons) {
                rows = matchedRowsTruncatedRefresh(oldRows: rows, newSeasons: seasons)
                syncNextSequenceFromRows()
                buildCount += 1
                return
            }
            resetAndBuild(from: seasons)
            return
        }
        if seasons.count == rows.count {
            if canSafelyReconcile(oldRows: rows, newSeasons: seasons) {
                rows = matchedRows(oldRows: rows, newSeasons: seasons)
                buildCount += 1
                return
            }
            resetAndBuild(from: seasons)
            return
        }
        // refresh 不应返回比现有累计页更长的列表；保守 reset。
        resetAndBuild(from: seasons)
    }

    private func reconcileLoadMore(with seasons: [OfficialCapitalRaidSeason]) {
        if rows.isEmpty {
            resetAndBuild(from: seasons)
            return
        }
        if seasons.count == rows.count {
            if positionalTripleMatch(rows, seasons) {
                rows = zip(rows, seasons).map { CapitalRaidSeasonRow(id: $0.id, season: $1) }
                return
            }
            resetAndBuild(from: seasons)
            return
        }
        guard seasons.count > rows.count else {
            resetAndBuild(from: seasons)
            return
        }
        let prefix = Array(seasons.prefix(rows.count))
        guard positionalTripleMatch(rows, prefix) else {
            resetAndBuild(from: seasons)
            return
        }
        var updated = rows
        for season in seasons[rows.count...] {
            updated.append(makeRow(for: season))
        }
        rows = updated
        buildCount += 1
    }

    private func resetAndBuild(from seasons: [OfficialCapitalRaidSeason]) {
        generation &+= 1
        if generation == 0 { generation = 1 }
        nextSequenceByTriple = [:]
        rows = seasons.map(makeRow(for:))
        buildCount += 1
    }

    private func makeRow(for season: OfficialCapitalRaidSeason) -> CapitalRaidSeasonRow {
        let tripleKey = CapitalRaidRowIdentity.tripleKey(for: season)
        let seq = nextSequenceByTriple[tripleKey, default: 0]
        nextSequenceByTriple[tripleKey] = seq + 1
        let id = "raid:g\(generation):\(tripleKey)#\(seq)"
        return CapitalRaidSeasonRow(id: id, season: season)
    }

    private func syncNextSequenceFromRows() {
        nextSequenceByTriple = [:]
        for row in rows {
            let tripleKey = CapitalRaidRowIdentity.tripleKey(for: row.season)
            guard let seq = parseSequence(from: row.id) else { continue }
            nextSequenceByTriple[tripleKey] = max(nextSequenceByTriple[tripleKey, default: 0], seq + 1)
        }
    }

    private func parseSequence(from id: String) -> Int? {
        guard let hashIndex = id.lastIndex(of: "#") else { return nil }
        let suffix = id[id.index(after: hashIndex)...]
        return Int(suffix)
    }

    private func positionalTripleMatch(
        _ oldRows: [CapitalRaidSeasonRow],
        _ newSeasons: [OfficialCapitalRaidSeason]
    ) -> Bool {
        guard oldRows.count == newSeasons.count else { return false }
        for (row, season) in zip(oldRows, newSeasons) {
            if CapitalRaidRowIdentity.tripleKey(for: row.season)
                != CapitalRaidRowIdentity.tripleKey(for: season) {
                return false
            }
        }
        return true
    }

    /// 可证明匹配时返回 new 顺序下的旧 row ID；歧义时 nil（fail-closed）。
    private func canSafelyReconcile(
        oldRows: [CapitalRaidSeasonRow],
        newSeasons: [OfficialCapitalRaidSeason]
    ) -> Bool {
        CapitalRaidSeasonMatcher.canSafelyMatch(
            oldSeasons: oldRows.map(\.season),
            newSeasons: newSeasons
        )
    }

    /// 截短 refresh：完整旧缓存参与 duplicate 歧义判断；仅唯一 exact anchor 可保留 ID。
    private func canSafelyReconcileTruncatedRefresh(
        oldRows: [CapitalRaidSeasonRow],
        newSeasons: [OfficialCapitalRaidSeason]
    ) -> Bool {
        CapitalRaidSeasonMatcher.matchTruncatedRefreshOldIndices(
            oldSeasons: oldRows.map(\.season),
            newSeasons: newSeasons
        ) != nil
    }

    private func matchedRows(
        oldRows: [CapitalRaidSeasonRow],
        newSeasons: [OfficialCapitalRaidSeason]
    ) -> [CapitalRaidSeasonRow] {
        guard let oldIndices = CapitalRaidSeasonMatcher.matchOldIndices(
            oldSeasons: oldRows.map(\.season),
            newSeasons: newSeasons
        ) else {
            return newSeasons.map { CapitalRaidSeasonRow(id: makeRow(for: $0).id, season: $0) }
        }
        return zip(oldIndices, newSeasons).map { CapitalRaidSeasonRow(id: oldRows[$0].id, season: $1) }
    }

    private func matchedRowsTruncatedRefresh(
        oldRows: [CapitalRaidSeasonRow],
        newSeasons: [OfficialCapitalRaidSeason]
    ) -> [CapitalRaidSeasonRow] {
        guard let oldIndices = CapitalRaidSeasonMatcher.matchTruncatedRefreshOldIndices(
            oldSeasons: oldRows.map(\.season),
            newSeasons: newSeasons
        ) else {
            return newSeasons.map { CapitalRaidSeasonRow(id: makeRow(for: $0).id, season: $0) }
        }
        return zip(oldIndices, newSeasons).map { CapitalRaidSeasonRow(id: oldRows[$0].id, season: $1) }
    }
}

File: packages/domain/src/official/capital-raid-row-cache.ts
import {
  canSafelyMatchCapitalRaidSeasons,
  matchCapitalRaidOldIndices,
  matchTruncatedCapitalRaidRefreshOldIndices,
} from './capital-raid-matcher';
import { capitalRaidTripleKey, type CapitalRaidSeasonRow } from './capital-raid-row-identity';
import type { OfficialCapitalRaidPage, OfficialCapitalRaidSeason } from './models/capital-raid';
import type { CapitalRaidLoadMoreReconciliation } from './pagination-logic';

/** 突袭周末 row identity 生命周期缓存（对齐 CapitalRaidRowCache.swift）。 */
export type CapitalRaidRowCacheUpdate =
  | { readonly kind: 'initial'; readonly page: OfficialCapitalRaidPage }
  | { readonly kind: 'refreshSuccess'; readonly page: OfficialCapitalRaidPage }
  | {
      readonly kind: 'loadMoreSuccess';
      readonly page: OfficialCapitalRaidPage;
      readonly reconciliation?: CapitalRaidLoadMoreReconciliation;
    }
  | { readonly kind: 'parserRebuild'; readonly page: OfficialCapitalRaidPage }
  | { readonly kind: 'failureRetain' }
  | { readonly kind: 'clear' };

export class CapitalRaidRowCache {
  private _generation = 0;
  private _rows: CapitalRaidSeasonRow[] = [];
  private _buildCount = 0;
  private nextSequenceByTriple = new Map<string, number>();

  get generation(): number {
    return this._generation;
  }

  get rows(): readonly CapitalRaidSeasonRow[] {
    return this._rows;
  }

  get buildCount(): number {
    return this._buildCount;
  }

  apply(update: CapitalRaidRowCacheUpdate): readonly CapitalRaidSeasonRow[] {
    switch (update.kind) {
      case 'failureRetain':
        return this._rows;
      case 'clear':
        this._generation = 0;
        this._rows = [];
        this.nextSequenceByTriple.clear();
        return this._rows;
      case 'initial':
      case 'parserRebuild':
        this.resetAndBuild(update.page.page.items);
        return this._rows;
      case 'refreshSuccess':
        this.reconcileRefresh(update.page.page.items);
        return this._rows;
      case 'loadMoreSuccess':
        if ((update.reconciliation ?? 'identityPreserving') === 'ambiguous') {
          this.resetAndBuild(update.page.page.items);
        } else {
          this.reconcileLoadMore(update.page.page.items);
        }
        return this._rows;
    }
  }

  private reconcileRefresh(seasons: readonly OfficialCapitalRaidSeason[]): void {
    if (this._rows.length === 0) {
      this.resetAndBuild(seasons);
      return;
    }
    if (seasons.length < this._rows.length) {
      if (seasons.length === 0) {
        this.resetAndBuild(seasons);
        return;
      }
      if (this.canSafelyReconcileTruncatedRefresh(seasons)) {
        this._rows = this.matchedRowsTruncatedRefresh(seasons);
        this.syncNextSequenceFromRows();
        this._buildCount += 1;
        return;
      }
      this.resetAndBuild(seasons);
      return;
    }
    if (seasons.length === this._rows.length) {
      if (this.canSafelyReconcile(seasons)) {
        this._rows = this.matchedRows(seasons);
        this._buildCount += 1;
        return;
      }
      this.resetAndBuild(seasons);
      return;
    }
    this.resetAndBuild(seasons);
  }

  private reconcileLoadMore(seasons: readonly OfficialCapitalRaidSeason[]): void {
    if (this._rows.length === 0) {
      this.resetAndBuild(seasons);
      return;
    }
    if (seasons.length === this._rows.length) {
      if (this.canSafelyReconcile(seasons)) {
        this._rows = this.matchedRows(seasons);
        this._buildCount += 1;
        return;
      }
      this.resetAndBuild(seasons);
      return;
    }
    if (seasons.length <= this._rows.length) {
      this.resetAndBuild(seasons);
      return;
    }
    const prefix = seasons.slice(0, this._rows.length);
    if (this.canSafelyReconcile(prefix)) {
      const updated = this.matchedRows(prefix);
      for (const season of seasons.slice(this._rows.length)) {
        updated.push(this.makeRow(season));
      }
      this._rows = updated;
      this._buildCount += 1;
      return;
    }
    this.resetAndBuild(seasons);
  }

  private resetAndBuild(seasons: readonly OfficialCapitalRaidSeason[]): void {
    this._generation += 1;
    if (this._generation === 0) {
      this._generation = 1;
    }
    this.nextSequenceByTriple.clear();
    this._rows = seasons.map((season) => this.makeRow(season));
    this._buildCount += 1;
  }

  private makeRow(season: OfficialCapitalRaidSeason): CapitalRaidSeasonRow {
    const tripleKey = capitalRaidTripleKey(season);
    const seq = this.nextSequenceByTriple.get(tripleKey) ?? 0;
    this.nextSequenceByTriple.set(tripleKey, seq + 1);
    const id = `raid:g${this._generation}:${tripleKey}#${seq}`;
    return { id, season };
  }

  private syncNextSequenceFromRows(): void {
    this.nextSequenceByTriple.clear();
    for (const row of this._rows) {
      const tripleKey = capitalRaidTripleKey(row.season);
      const seq = parseSequenceFromId(row.id);
      if (seq === undefined) {
        continue;
      }
      this.nextSequenceByTriple.set(
        tripleKey,
        Math.max(this.nextSequenceByTriple.get(tripleKey) ?? 0, seq + 1),
      );
    }
  }

  private canSafelyReconcile(newSeasons: readonly OfficialCapitalRaidSeason[]): boolean {
    return canSafelyMatchCapitalRaidSeasons(
      this._rows.map((row) => row.season),
      newSeasons,
    );
  }

  private canSafelyReconcileTruncatedRefresh(
    newSeasons: readonly OfficialCapitalRaidSeason[],
  ): boolean {
    return (
      matchTruncatedCapitalRaidRefreshOldIndices(
        this._rows.map((row) => row.season),
        newSeasons,
      ) !== undefined
    );
  }

  private matchedRows(newSeasons: readonly OfficialCapitalRaidSeason[]): CapitalRaidSeasonRow[] {
    const oldIndices = matchCapitalRaidOldIndices(
      this._rows.map((row) => row.season),
      newSeasons,
    );
    if (oldIndices === undefined) {
      return newSeasons.map((season) => this.makeRow(season));
    }
    return oldIndices.map((oldIndex, newIndex) => ({
      id: this._rows[oldIndex]!.id,
      season: newSeasons[newIndex]!,
    }));
  }

  private matchedRowsTruncatedRefresh(
    newSeasons: readonly OfficialCapitalRaidSeason[],
  ): CapitalRaidSeasonRow[] {
    const oldIndices = matchTruncatedCapitalRaidRefreshOldIndices(
      this._rows.map((row) => row.season),
      newSeasons,
    );
    if (oldIndices === undefined) {
      return newSeasons.map((season) => this.makeRow(season));
    }
    return oldIndices.map((oldIndex, newIndex) => ({
      id: this._rows[oldIndex]!.id,
      season: newSeasons[newIndex]!,
    }));
  }
}

function parseSequenceFromId(id: string): number | undefined {
  const hashIndex = id.lastIndexOf('#');
  if (hashIndex < 0) {
    return undefined;
  }
  const suffix = id.slice(hashIndex + 1);
  const parsed = Number.parseInt(suffix, 10);
  return Number.isInteger(parsed) ? parsed : undefined;
}
